core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/AvroUpsertMonitor.scala

```diff
@@ -0,0 +1,237 @@
+             avroUpsertBucketName: String,
+             batchSize: Int)(implicit cs: ContextShift[IO]): Props =
+    Props(new AvroUpsertMonitorActor(pollInterval, pollIntervalJitter, workspaceService, googleServicesDAO, samDAO, googleStorage, pubSubDao, pubSubSubscriptionName, avroUpsertBucketName, batchSize))
+}
+
+class AvroUpsertMonitorActor(
+                                  val pollInterval: FiniteDuration,
+                                  pollIntervalJitter: FiniteDuration,
+                                  workspaceService: UserInfo => WorkspaceService,
+                                  googleServicesDAO: GoogleServicesDAO,
+                                  samDAO: SamDAO,
+                                  googleStorage: GoogleStorageService[IO],
+                                  pubSubDao: GooglePubSubDAO,
+                                  pubSubSubscriptionName: String,
+                                  avroUpsertBucketName: String,
+                                  batchSize: Int)(implicit cs: ContextShift[IO])
+  extends Actor
+    with LazyLogging
+    with FutureSupport {
+  import AvroUpsertMonitor._
+  import context._
+
+  case class AvroMetadataJson(namespace: String, name: String, userSubjectId: String, userEmail: String)
+  implicit val avroMetadataJsonFormat = jsonFormat4(AvroMetadataJson)
+
+  self ! StartMonitorPass
+
+  // fail safe in case this actor is idle too long but not too fast (1 second lower limit)
+  setReceiveTimeout(max((pollInterval + pollIntervalJitter) * 10, 1 second))
+
+  private def max(durations: FiniteDuration*): FiniteDuration = {
+    implicit val finiteDurationIsOrdered = scala.concurrent.duration.FiniteDuration.FiniteDurationIsOrdered
+    durations.max
+  }
+
+  override def receive = {
+    case StartMonitorPass =>
+      // start the process by pulling a message and sending it back to self
+      pubSubDao.pullMessages(pubSubSubscriptionName, 1).map(_.headOption) pipeTo self
+
+    case Some(message: PubSubMessage) =>
+      logger.info(s"received avro upsert message: $message")
+      val (jobId, file) = parseMessage(message)
+
+      file match {
+        case "upsert.json" => initAvroUpsert(jobId, file).map(response => (response, message.ackId)) pipeTo self
+        case _ => acknowledgeMessage(message.ackId) pipeTo self //some other file (i.e. success/failure log, metadata.json)
+      }
+
+    case None =>
+      // there was no message so wait and try again
+      val nextTime = org.broadinstitute.dsde.workbench.util.addJitter(pollInterval, pollIntervalJitter)
+      system.scheduler.scheduleOnce(nextTime.asInstanceOf[FiniteDuration], self, StartMonitorPass)
+
+    case ((), ackId: String) => self ! None
+      acknowledgeMessage(ackId).map(_ => StartMonitorPass) pipeTo self
+//
+    case Status.Failure(t) => throw t
+
+    case ReceiveTimeout =>
+      throw new WorkbenchException("AvroUpsertMonitorActor has received no messages for too long")
```

Comment:
Is this an escape hatch in case the actor gets stuck for some unknown reason? Does throwing an exception here cause the actor to be replaced (according to the supervisor strategy above)? Does the time spent in `initAvroUpsert` count as activity?

Reply:
Yes exactly it's an escape hatch. 

And yes it seems to count because the timeout would be around 10-12 minutes and I've had an initAvroUpsert last as long as 14 mins in the case of a very low batchSize.

Will double check with a very low timeout, though.